controllers/metallb_controller.go

```diff
@@ -160,91 +161,62 @@ func (r *MetalLBReconciler) SetupWithManager(mgr ctrl.Manager, bgpType string) e
+	// TODO: patch rbac proxy settings if needed.
+	client := action.NewInstall(actionConfig)
+	client.ReleaseName = helmReleaseName
+	client.Namespace = r.HelmEnvSettings.Namespace()
+	client.DryRun = true
+	client.Replace = true
+	client.ClientOnly = true
+	chartPath, err := client.ChartPathOptions.LocateChart(ManifestPath, r.HelmEnvSettings)
 	if err != nil {
 		return err
 	}
-	data.Data["MLBindPort"], err = valueWithDefault("MEMBER_LIST_BIND_PORT", 7946)
+	chart, err := loader.Load(chartPath)
 	if err != nil {
+		logger.Error(err, "metallb chart loading failed")
 		return err
 	}
-
-	objs, err := render.RenderDir(ManifestPath, &data)
+	release, err := client.Run(chart, chartValues)
 	if err != nil {
-		logger.Error(err, "Fail to render config daemon manifests")
+		logger.Error(err, "failed to generate manifests from metallb chart")
 		return err
 	}
-
-	// We shouldn't spam the api server trying to apply PodMonitors if the resource isn't installed.
-	deployPodMonitors := os.Getenv("DEPLOY_PODMONITORS") == "true"
-	if podMonitorAvailable(r.Client) && deployPodMonitors {
-		podmonitors, err := render.RenderDir(PodMonitorsPath, &data)
-		if err != nil {
-			logger.Error(err, "Fail to render PodMonitors manifests")
-			return err
-		}
-		objs = append(objs, podmonitors...)
-	} else {
-		logger.Info("PodMonitors Resource not available in the cluster. Will not try to apply them.")
+	objs, err := render.ParseRenderedManifest(release.Manifest)
+	if err != nil {
+		logger.Error(err, "fail to parse rendered metallb manifests")
+		return err
 	}
 
 	for _, obj := range objs {
-		objNS := obj.GetNamespace()
-		if objNS != "" { // Avoid setting reference on a cluster-scoped resource.
-			if err := controllerutil.SetControllerReference(config, obj, r.Scheme); err != nil {
-				return errors.Wrapf(err, "Failed to set controller reference to %s %s", obj.GetNamespace(), obj.GetName())
-			}
+		objKind := obj.GetKind()
+		// Skip applying role and role binding object
+		if objKind == "Role" || objKind == "RoleBinding" {
+			continue
 		}
-		if obj.GetKind() == "DaemonSet" &&
-			(len(config.Spec.SpeakerNodeSelector) > 0 || len(config.Spec.SpeakerTolerations) > 0) {
-			scheme := kscheme.Scheme
-			ds := &appsv1.DaemonSet{}
-			err = scheme.Convert(obj, ds, nil)
-			if err != nil {
-				logger.Error(err, "Fail to convert MetalLB object to DaemonSet")
-				return err
-			}
-			if len(config.Spec.SpeakerNodeSelector) > 0 {
-				ds.Spec.Template.Spec.NodeSelector = config.Spec.SpeakerNodeSelector
-			}
-			if len(config.Spec.SpeakerTolerations) > 0 {
-				ds.Spec.Template.Spec.Tolerations = config.Spec.SpeakerTolerations
-			}
-			err = scheme.Convert(ds, obj, nil)
-			if err != nil {
-				logger.Error(err, "Fail to convert DaemonSet to MetalLB object")
-				return err
+		// Set namespace and controller reference on object (daemon set, deployment or config map)
+		if objKind != "PodSecurityPolicy" {
```

Comment:
The old logic wasn't working?

objNS := obj.GetNamespace()
		if objNS != "" { // Avoid setting reference on a cluster-scoped resource.

Seems more future proof.

Reply:
yes, namespace is not on the manifests, so needed this to set the namespace other than psp objects.

Reply:
But we are setting the namespace here https://github.com/metallb/metallb-operator/blob/7c8222e6377951a4246d12a78a264b366ff81288/pkg/helm/helm.go#L64
so maybe we can just pass the right namespace there?

Reply:
tried setting namespace via `helmEnvSettings`, but didn't work, looks like this is useful only for `helm install `case.